File: sdks/python/src/a11yfy/webhooks.py
# Hand-written overlay — NOT generated. Protected by .fernignore.
#
# Webhook signature verification (Stripe constructEvent pattern), matched to
# the a11yfy contract (API-CONTRACT.md §5.3):
#   header:  X-A11yfy-Signature: ts=<unix_seconds>;h1=<hex_hmac>
#   signed:  "{ts}:{raw_request_body}"        <-- COLON separator!
#   algo:    HMAC-SHA256, hex, constant-time compare, 300s replay window

from __future__ import annotations

import hashlib
import hmac
import json
import time
import typing

__all__ = ["Webhook", "WebhookEvent", "WebhookVerificationError"]

#: Replay-protection window in seconds (contract §5.3).
DEFAULT_TOLERANCE = 300


class WebhookVerificationError(Exception):
    """The webhook signature could not be verified."""


class WebhookEvent:
    """Verified webhook payload (terminal job notification).

    Attributes mirror the contract payload (§5.2): job_id, status
    ('done' | 'failed' | 'partial'), credits_used, output_url (on success),
    error (on failure), completed_at.
    """

    def __init__(self, data: typing.Dict[str, typing.Any]) -> None:
        self._data = data
        self.job_id: str = data.get("job_id", "")
        self.status: str = data.get("status", "")
        self.credits_used: typing.Optional[int] = data.get("credits_used")
        self.output_url: typing.Optional[str] = data.get("output_url")
        self.error: typing.Optional[str] = data.get("error")
        self.completed_at: typing.Optional[str] = data.get("completed_at")

    @property
    def is_success(self) -> bool:
        return self.status == "done"

    def to_dict(self) -> typing.Dict[str, typing.Any]:
        return dict(self._data)

    def __repr__(self) -> str:  # pragma: no cover
        return f"WebhookEvent(job_id={self.job_id!r}, status={self.status!r})"


def _parse_header(header: str) -> typing.Tuple[int, str]:
    ts: typing.Optional[str] = None
    h1: typing.Optional[str] = None
    for part in header.split(";"):
        key, _, value = part.strip().partition("=")
        if key == "ts":
            ts = value
        elif key == "h1":
            h1 = value
    if ts is None or h1 is None or not ts.isdigit() or not h1:
        raise WebhookVerificationError(
            "Unable to extract ts and h1 from the X-A11yfy-Signature header."
        )
    return int(ts), h1


class Webhook:
    """Verify and parse a11yfy webhook deliveries.

    Usage (e.g. in a Flask/FastAPI handler):
        event = Webhook.construct_event(
            payload=request.body,                              # RAW bytes!
            signature_header=request.headers["X-A11yfy-Signature"],
            secret=os.environ["A11YFY_WEBHOOK_SECRET"],
        )
        if event.is_success:
            download(event.output_url)
    """

    @staticmethod
    def construct_event(
        payload: typing.Union[bytes, str],
        signature_header: str,
        secret: str,
        tolerance: int = DEFAULT_TOLERANCE,
    ) -> WebhookEvent:
        raw = payload.encode("utf-8") if isinstance(payload, str) else payload
        ts, h1 = _parse_header(signature_header)

        if tolerance and abs(time.time() - ts) > tolerance:
            raise WebhookVerificationError(
                f"Timestamp outside the {tolerance}s tolerance zone."
            )

        signed_payload = str(ts).encode("ascii") + b":" + raw
        expected = hmac.new(
            secret.encode("utf-8"), signed_payload, hashlib.sha256
        ).hexdigest()
        if not hmac.compare_digest(expected, h1):
            raise WebhookVerificationError(
                "Signature mismatch — payload was not signed with this secret."
            )

        try:
            data = json.loads(raw.decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError) as exc:
            raise WebhookVerificationError("Payload is not valid JSON.") from exc
        return WebhookEvent(data)

File: sdks/typescript/src/wrapper/webhooks.ts
// Hand-written overlay — NOT generated. Protected by .fernignore.
//
// Webhook signature verification (Stripe constructEvent pattern), matched to
// the a11yfy contract (API-CONTRACT.md §5.3):
//   header:  X-A11yfy-Signature: ts=<unix_seconds>;h1=<hex_hmac>
//   signed:  "{ts}:{raw_request_body}"        <-- COLON separator!
//   algo:    HMAC-SHA256, hex, constant-time compare, 300s replay window

import { createHmac, timingSafeEqual } from "node:crypto";

/** Replay-protection window in seconds (contract §5.3). */
export const DEFAULT_TOLERANCE_SECONDS = 300;

/** The webhook signature could not be verified. */
export class WebhookVerificationError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "WebhookVerificationError";
    }
}

/** Verified webhook payload (terminal job notification, contract §5.2). */
export interface WebhookEvent {
    job_id: string;
    status: "done" | "failed" | "partial";
    credits_used?: number;
    output_url?: string;
    error?: string;
    completed_at?: string;
}

function parseHeader(header: string): { ts: number; h1: string } {
    let ts: string | undefined;
    let h1: string | undefined;
    for (const part of header.split(";")) {
        const idx = part.indexOf("=");
        if (idx <= 0) continue;
        const key = part.slice(0, idx).trim();
        const value = part.slice(idx + 1).trim();
        if (key === "ts") ts = value;
        else if (key === "h1") h1 = value;
    }
    if (!ts || !h1 || !/^\d+$/.test(ts)) {
        throw new WebhookVerificationError(
            "Unable to extract ts and h1 from the X-A11yfy-Signature header.",
        );
    }
    return { ts: Number(ts), h1 };
}

/**
 * Verify and parse an a11yfy webhook delivery.
 *
 * @example
 * import { Webhooks } from "@a11yfy/sdk";
 *
 * const event = Webhooks.constructEvent(
 *     rawBody,                                   // RAW string/Buffer!
 *     req.headers["x-a11yfy-signature"] as string,
 *     process.env.A11YFY_WEBHOOK_SECRET!,
 * );
 * if (event.status === "done") download(event.output_url!);
 */
export const Webhooks = {
    constructEvent(
        payload: string | Buffer,
        signatureHeader: string,
        secret: string,
        toleranceSeconds: number = DEFAULT_TOLERANCE_SECONDS,
    ): WebhookEvent {
        const raw = typeof payload === "string" ? Buffer.from(payload, "utf-8") : payload;
        const { ts, h1 } = parseHeader(signatureHeader);

        if (toleranceSeconds && Math.abs(Date.now() / 1000 - ts) > toleranceSeconds) {
            throw new WebhookVerificationError(
                `Timestamp outside the ${toleranceSeconds}s tolerance zone.`,
            );
        }

        const signedPayload = Buffer.concat([Buffer.from(`${ts}:`, "ascii"), raw]);
        const expected = createHmac("sha256", secret).update(signedPayload).digest("hex");
        const a = Buffer.from(expected, "utf-8");
        const b = Buffer.from(h1, "utf-8");
        if (a.length !== b.length || !timingSafeEqual(a, b)) {
            throw new WebhookVerificationError(
                "Signature mismatch — payload was not signed with this secret.",
            );
        }

        try {
            return JSON.parse(raw.toString("utf-8")) as WebhookEvent;
        } catch {
            throw new WebhookVerificationError("Payload is not valid JSON.");
        }
    },
};
